pol-core/ecompile/EfswFileWatchListener.cpp

```diff
@@ -0,0 +1,175 @@
+#include "EfswFileWatchListener.h"
+
+#include "clib/logfacility.h"
+#include <filesystem>
+#include <string>
+
+namespace fs = std::filesystem;
+
+namespace Pol::ECompile
+{
+EfswFileWatchListener::EfswFileWatchListener(
+    efsw::FileWatcher& watcher, const std::set<std::filesystem::path>& extension_filter )
+    : watcher( watcher ),
+      extension_filter( extension_filter ),
+      watched_dirs(),
+      watched_files(),
+      mutex(),
+      messages(),
+      handle_messages_by( 0 )
+{
+}
+
+EfswFileWatchListener::~EfswFileWatchListener()
+{
+  for ( const auto& v : dir_to_watchid )
+  {
+    watcher.removeWatch( v.second );
+  }
+}
+
+// Implements the libefsw virtual function for handling filesystem events.
+void EfswFileWatchListener::handleFileAction( efsw::WatchID /*watchid*/, const std::string& dir,
+                                              const std::string& filename, efsw::Action action,
+                                              std::string oldFilename )
+{
+  fs::path dirPath( dir );
+  auto filepath = dirPath / fs::path( filename );
+
+  // Only send events for watched files
+  if ( !is_watched( filepath ) )
+  {
+    return;
+  }
+
+  switch ( action )
+  {
+  case efsw::Actions::Delete:
+    // Delete message: Empty filepath, Non-empty old_filepath
+    add_message( WatchFileMessage{ fs::path( "" ), filepath } );
+    break;
+
+  case efsw::Actions::Moved:
+  {
+    // Moved message: Non-empty filepath, Non-empty old_filepath
+    auto oldFilepath = dirPath / fs::path( oldFilename );
+    add_message( WatchFileMessage{ filepath, oldFilepath } );
+    break;
+  }
+  case efsw::Actions::Modified:
+  case efsw::Actions::Add:
+    // Changed message: Non-empty filepath, empty old_filepath
+    add_message( WatchFileMessage{ filepath, fs::path( "" ) } );
+    break;
+
+  default:
+    break;
+  }
+}
+
+// Adds a file to watched files. Returns `true` if the file was newly added to
+// `watched_files`.
+bool EfswFileWatchListener::add_file( const fs::path& filepath )
```

Comment:
const ref plus .emplace still gives you a copy. Have plain fs::path + std::move for emplace removes the extra copy